sematic/abstract_future.py

```diff
@@ -118,24 +162,32 @@ def __init__(
     ):
         self.id: str = make_future_id()
         self.calculator = calculator
-        self.kwargs = kwargs
-        # We don't want to replace futures in kwargs, because it holds
-        # the source of truth for the future graph. Instead we have concrete
-        # values in resolved_kwargs
-        # It will be set only once all input values are resolved
-        self.resolved_kwargs: Dict[str, Any] = {}
         self.value: Any = None
         self.state: FutureState = FutureState.CREATED
         self.parent_future: Optional["AbstractFuture"] = None
         self.nested_future: Optional["AbstractFuture"] = None
 
+        try:
+            external_resources: List[ExternalResource] = list(
+                context().private.external_resources  # type: ignore
+            )
+        except NotInSematicFuncError:
+            # the root future is not created inside another func
+            external_resources = []
```

Comment:
```suggestion
        external_resources: List[ExternalResource] = []
        try:
            external_resources: List[ExternalResource] = list(
                context().private.external_resources  # type: ignore
            )
        except NotInSematicFuncError:
            # the root future is not created inside another func
            pass
```
This conserves the typing info even if `NotInSematicFuncError` is raised.

Also, what is that `type: ignore` for?